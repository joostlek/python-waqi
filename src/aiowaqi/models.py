"""Asynchronous Python client for the WAQI API."""
from __future__ import annotations

from contextlib import suppress
from dataclasses import dataclass
from datetime import datetime
from enum import StrEnum
from typing import Any, Self

from aiowaqi.util import to_nullable_enum


class Pollutant(StrEnum):
    """Enum of pollutants."""

    CARBON_MONOXIDE = "co"
    NITROGEN_DIOXIDE = "no2"
    OZONE = "o3"
    SULFUR_DIOXIDE = "so2"
    PM10 = "pm10"
    PM25 = "pm25"
    NEPHELOMETRY = "neph"


@dataclass(slots=True)
class Attribution:
    """Represents an attribution."""

    url: str
    name: str
    logo: str | None

    @classmethod
    def from_dict(cls, attribution: dict[str, Any]) -> Self:
        """Initialize from a dict."""
        return cls(
            url=attribution["url"],
            name=attribution["name"].strip(),
            logo=attribution.get("logo"),
        )


@dataclass(slots=True)
class Coordinates:
    """Represents coordinates."""

    latitude: float
    longitude: float


@dataclass(slots=True)
class Location:
    """Represents a location object."""

    external_url: str
    name: str
    coordinates: Coordinates

    @classmethod
    def from_dict(cls, location: dict[str, Any]) -> Self:
        """Initialize from a dict."""
        return cls(
            external_url=location["url"],
            name=location["name"],
            coordinates=Coordinates(
                latitude=location["geo"][0],
                longitude=location["geo"][1],
            ),
        )


@dataclass(slots=True)
class City(Location):
    """Represents a city object."""

    location: str | None

    @classmethod
    def from_dict(cls, location: dict[str, Any]) -> Self:
        """Initialize from a dict."""
        loc = location["location"]
        if not loc:
            loc = None
        return cls(
            external_url=location["url"],
            name=location["name"],
            coordinates=Coordinates(
                latitude=location["geo"][0],
                longitude=location["geo"][1],
            ),
            location=loc,
        )


@dataclass(slots=True)
class Station(Location):
    """Represents a station object."""


@dataclass(slots=True)
class WAQIExtendedAirQuality:
    """Represents extended air quality data."""

    carbon_monoxide: float | None
    humidity: float | None
    nephelometry: float | None
    nitrogen_dioxide: float | None
    ozone: float | None
    pressure: float | None
    sulfur_dioxide: float | None
    pm10: float | None
    pm25: float | None
    temperature: float | None

    @classmethod
    def from_dict(cls, air_quality: dict[str, Any]) -> Self:
        """Initialize from a dict."""
        return cls(
            carbon_monoxide=air_quality.get("co", {}).get("v"),
            humidity=air_quality.get("h", {}).get("v"),
            nephelometry=air_quality.get("neph", {}).get("v"),
            nitrogen_dioxide=air_quality.get("no2", {}).get("v"),
            ozone=air_quality.get("o3", {}).get("v"),
            pressure=air_quality.get("p", {}).get("v"),
            sulfur_dioxide=air_quality.get("so2", {}).get("v"),
            pm10=air_quality.get("pm10", {}).get("v"),
            pm25=air_quality.get("pm25", {}).get("v"),
            temperature=air_quality.get("t", {}).get("v"),
        )


@dataclass(slots=True)
class WAQIAirQuality:
    """Represents the air quality data from WAQI."""

    air_quality_index: int | None
    station_id: int
    attributions: list[Attribution]
    city: City
    extended_air_quality: WAQIExtendedAirQuality
    dominant_pollutant: Pollutant | None
    measured_at: datetime | None

    @classmethod
    def from_dict(cls, air_quality: dict[str, Any]) -> Self:
        """Initialize from a dict."""
        aqi = None
        with suppress(ValueError):
            aqi = int(air_quality["aqi"])

        dominant_pollutant = air_quality["dominentpol"]
        if dominant_pollutant == "":
            dominant_pollutant = None
        else:
            dominant_pollutant = to_nullable_enum(Pollutant, air_quality["dominentpol"])

        measured_at: datetime | None = None
        if "iso" in air_quality["time"]:
            measured_at = datetime.fromisoformat(air_quality["time"]["iso"])

        return cls(
            air_quality_index=aqi,
            station_id=air_quality["idx"],
            attributions=[
                Attribution.from_dict(attribution)
                for attribution in air_quality["attributions"]
            ],
            city=City.from_dict(air_quality["city"]),
            extended_air_quality=WAQIExtendedAirQuality.from_dict(air_quality["iaqi"]),
            dominant_pollutant=dominant_pollutant,
            measured_at=measured_at,
        )


@dataclass(slots=True)
class WAQISearchResult:
    """Represents a search result from the WAQI api."""

    air_quality_index: int | None
    station_id: int
    station: Station

    @classmethod
    def from_dict(cls, result: dict[str, Any]) -> Self:
        """Initialize from a dict."""
        aqi = None
        with suppress(ValueError):
            aqi = int(result["aqi"])

        return cls(
            air_quality_index=aqi,
            station_id=result["uid"],
            station=Station.from_dict(result["station"]),
        )
