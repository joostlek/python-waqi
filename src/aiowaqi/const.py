"""Asynchronous Python client for the WAQI API."""
from __future__ import annotations

import logging

LOGGER = logging.getLogger(__package__)
