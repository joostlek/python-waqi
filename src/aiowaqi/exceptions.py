"""Asynchronous Python client for the WAQI API."""


class WAQIError(Exception):
    """Generic exception."""


class WAQIConnectionError(WAQIError):
    """WAQI connection exception."""


class WAQIUnknownCityError(WAQIError):
    """WAQI unknown city exception."""


class WAQIUnknownStationError(WAQIError):
    """WAQI unknown station exception."""


class WAQIAuthenticationError(WAQIError):
    """WAQI authentication exception."""
