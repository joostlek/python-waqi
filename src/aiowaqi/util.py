"""Asynchronous Python client for the WAQI API."""
from __future__ import annotations

from typing import Any, TypeVar

from .const import LOGGER

_EnumT = TypeVar("_EnumT")


def to_nullable_enum(
    enum_class: type[_EnumT],
    value: Any,
) -> _EnumT | None:
    """Convert a value to an enum and log if it doesn't exist."""
    try:
        return enum_class(value)  # type: ignore[call-arg]
    except ValueError:
        LOGGER.warning(
            "%r is an unsupported value for %s, please report this at https://github.com/joostlek/python-waqi/issues",
            value,
            str(enum_class),
        )
        return None
