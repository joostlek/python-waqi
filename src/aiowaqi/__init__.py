"""Asynchronous Python client for the WAQI API."""

from .exceptions import (
    WAQIAuthenticationError,
    WAQIConnectionError,
    WAQIError,
    WAQIUnknownCityError,
    WAQIUnknownStationError,
)
from .models import (
    Attribution,
    City,
    Coordinates,
    Location,
    WAQIAirQuality,
    WAQIExtendedAirQuality,
    WAQISearchResult,
)
from .waqi import WAQIClient

__all__ = [
    "WAQIError",
    "WAQIUnknownStationError",
    "WAQIUnknownCityError",
    "Location",
    "WAQISearchResult",
    "WAQIAuthenticationError",
    "WAQIClient",
    "WAQIConnectionError",
    "Attribution",
    "City",
    "Coordinates",
    "WAQIExtendedAirQuality",
    "WAQIAirQuality",
]
