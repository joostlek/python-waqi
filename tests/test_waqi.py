"""Tests for the WAQI Library."""
from __future__ import annotations

import asyncio
from dataclasses import asdict

import aiohttp
from aiohttp.web_request import BaseRequest
from aresponses import Response, ResponsesMockServer
import pytest
from syrupy import SnapshotAssertion

from aiowaqi import (
    WAQIAirQuality,
    WAQIAuthenticationError,
    WAQIClient,
    WAQIConnectionError,
    WAQIError,
    WAQISearchResult,
    WAQIUnknownStationError,
)

from . import load_fixture

WAQI_URL = "api.waqi.info"


@pytest.mark.parametrize(
    "city",
    [
        "utrecht",
        "maarssen",
        "klundert",
        "olivias",
        "failing_klundert",
    ],
)
async def test_by_city(
    aresponses: ResponsesMockServer,
    city: str,
    snapshot: SnapshotAssertion,
) -> None:
    """Test retrieving air quality by city."""
    aresponses.add(
        WAQI_URL,
        f"/feed/{city}?token=test",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixture(f"city_feed_{city}.json"),
        ),
        match_querystring=True,
    )
    async with aiohttp.ClientSession() as session:
        waqi = WAQIClient(session=session)
        waqi.authenticate("test")
        response: WAQIAirQuality = await waqi.get_by_city(city)
        assert asdict(response) == snapshot
        await waqi.close()


async def test_new_dominant_pol(
    aresponses: ResponsesMockServer,
    caplog: pytest.LogCaptureFixture,
) -> None:
    """Test retrieving new dominant pol."""
    aresponses.add(
        WAQI_URL,
        "/feed/maarssen?token=test",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixture("city_feed_new_dominant_pol.json"),
        ),
        match_querystring=True,
    )
    async with aiohttp.ClientSession() as session:
        waqi = WAQIClient(session=session)
        waqi.authenticate("test")
        await waqi.get_by_city("maarssen")
        assert (
            "'h' is an unsupported value for <enum 'Pollutant'>,"
            " please report this at https://github.com/joostlek/python-waqi/issues"
            in caplog.text
        )
        await waqi.close()


async def test_unknown_dominant_pol(
    aresponses: ResponsesMockServer,
) -> None:
    """Test retrieving unknown dominant pol."""
    aresponses.add(
        WAQI_URL,
        "/feed/maarssen?token=test",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixture("city_feed_unknown_dominant_pol.json"),
        ),
        match_querystring=True,
    )
    async with aiohttp.ClientSession() as session:
        waqi = WAQIClient(session=session)
        waqi.authenticate("test")
        air_quality = await waqi.get_by_city("maarssen")
        assert air_quality.dominant_pollutant is None
        await waqi.close()


async def test_own_session(
    aresponses: ResponsesMockServer,
) -> None:
    """Test creating own session."""
    aresponses.add(
        WAQI_URL,
        "/feed/utrecht",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixture("city_feed_utrecht.json"),
        ),
    )
    async with WAQIClient() as waqi:
        waqi.authenticate("test")
        await waqi.get_by_city("utrecht")
        assert waqi.session is not None


async def test_unexpected_server_response(
    aresponses: ResponsesMockServer,
) -> None:
    """Test handling unexpected response."""
    aresponses.add(
        WAQI_URL,
        "/feed/utrecht",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "plain/text"},
            text="Yes",
        ),
    )
    async with WAQIClient() as waqi:
        waqi.authenticate("test")
        with pytest.raises(WAQIError):
            assert await waqi.get_by_city("utrecht")


async def test_unknown_city(
    aresponses: ResponsesMockServer,
) -> None:
    """Test unknown city."""
    aresponses.add(
        WAQI_URL,
        "/feed/unknown",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixture("city_feed_unknown.json"),
        ),
    )
    async with WAQIClient() as waqi:
        waqi.authenticate("test")
        with pytest.raises(WAQIError):
            assert await waqi.get_by_city("unknown")


async def test_unauthenticated(
    aresponses: ResponsesMockServer,
) -> None:
    """Test unauthenticated."""
    aresponses.add(
        WAQI_URL,
        "/feed/utrecht",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixture("unauthenticated.json"),
        ),
    )
    async with WAQIClient() as waqi:
        waqi.authenticate("test")
        with pytest.raises(WAQIAuthenticationError):
            assert await waqi.get_by_city("utrecht")


async def test_timeout(aresponses: ResponsesMockServer) -> None:
    """Test request timeout."""

    # Faking a timeout by sleeping
    async def response_handler(_: BaseRequest) -> Response:
        """Response handler for this test."""
        await asyncio.sleep(2)
        return aresponses.Response(body="Goodmorning!")

    aresponses.add(
        WAQI_URL,
        "/feed/utrecht",
        "GET",
        response_handler,
    )
    async with aiohttp.ClientSession() as session:
        waqi = WAQIClient(session=session, request_timeout=1)
        waqi.authenticate("test")
        with pytest.raises(WAQIConnectionError):
            assert await waqi.get_by_city("utrecht")
        await waqi.close()


@pytest.mark.parametrize(
    "keyword",
    [
        "klundert",
        "failing_klundert",
        "unknown",
    ],
)
async def test_search(
    aresponses: ResponsesMockServer,
    keyword: str,
    snapshot: SnapshotAssertion,
) -> None:
    """Test searching stations."""
    aresponses.add(
        WAQI_URL,
        f"/search/?keyword={keyword}&token=test",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixture(f"search_{keyword}.json"),
        ),
        match_querystring=True,
    )
    async with WAQIClient() as waqi:
        waqi.authenticate("test")
        response: list[WAQISearchResult] = await waqi.search(keyword)
        res = [asdict(wsr) for wsr in response]
        assert res == snapshot


@pytest.mark.parametrize(
    "name",
    [
        "klundert",
        "failing_klundert",
    ],
)
async def test_get_by_name(
    aresponses: ResponsesMockServer,
    name: str,
    snapshot: SnapshotAssertion,
) -> None:
    """Test getting stations by name."""
    aresponses.add(
        WAQI_URL,
        f"/feed/{name}?token=test",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixture(f"name_feed_{name}.json"),
        ),
        match_querystring=True,
    )
    async with WAQIClient() as waqi:
        waqi.authenticate("test")
        response = await waqi.get_by_name(name)
        assert asdict(response) == snapshot


async def test_get_unknown_by_name(
    aresponses: ResponsesMockServer,
) -> None:
    """Test getting unknown station by name."""
    aresponses.add(
        WAQI_URL,
        "/feed/unknown",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixture("name_feed_unknown.json"),
        ),
    )
    async with WAQIClient() as waqi:
        waqi.authenticate("test")
        with pytest.raises(WAQIUnknownStationError):
            await waqi.get_by_name("unknown")


@pytest.mark.parametrize(
    "station_number",
    [
        6337,
        372382,
        10142,
        10002,
    ],
)
async def test_get_by_station_number(
    aresponses: ResponsesMockServer,
    station_number: int,
    snapshot: SnapshotAssertion,
) -> None:
    """Test getting stations by station_number."""
    aresponses.add(
        WAQI_URL,
        f"/feed/@{station_number}?token=test",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixture(f"station_number_feed_{station_number}.json"),
        ),
        match_querystring=True,
    )
    async with WAQIClient() as waqi:
        waqi.authenticate("test")
        response = await waqi.get_by_station_number(station_number)
        assert asdict(response) == snapshot


@pytest.mark.parametrize(
    "identifier",
    [
        "unknown",
        "@123946",
        "A10142",
    ],
)
async def test_get_unknown_by_station_number(
    aresponses: ResponsesMockServer,
    identifier: str,
) -> None:
    """Test getting unknown station by station_number."""
    aresponses.add(
        WAQI_URL,
        "/feed/@0",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixture(f"station_number_feed_{identifier}.json"),
        ),
    )
    async with WAQIClient() as waqi:
        waqi.authenticate("test")
        with pytest.raises(WAQIUnknownStationError):
            await waqi.get_by_station_number(0)


async def test_get_by_coordinates(
    aresponses: ResponsesMockServer,
    snapshot: SnapshotAssertion,
) -> None:
    """Test getting measuring station via coordinates."""
    aresponses.add(
        WAQI_URL,
        "/feed/geo:52.105031;5.124464",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixture("coordinates.json"),
        ),
    )
    async with WAQIClient() as waqi:
        waqi.authenticate("test")
        response = await waqi.get_by_coordinates(52.105031, 5.124464)
        assert asdict(response) == snapshot


async def test_get_by_ip(
    aresponses: ResponsesMockServer,
    snapshot: SnapshotAssertion,
) -> None:
    """Test getting measuring station via ip."""
    aresponses.add(
        WAQI_URL,
        "/feed/here",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixture("here.json"),
        ),
    )
    async with WAQIClient() as waqi:
        waqi.authenticate("test")
        response = await waqi.get_by_ip()
        assert asdict(response) == snapshot
