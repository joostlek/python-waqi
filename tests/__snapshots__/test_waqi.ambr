# serializer version: 1
# name: test_by_city[failing_klundert]
  dict({
    'air_quality_index': None,
    'attributions': list([
      dict({
        'logo': 'Netherland-RIVM.png',
        'name': 'RIVM - Rijksinstituut voor Volksgezondheid en Milieum, Landelijk Meetnet Luchtkwaliteit',
        'url': 'http://www.luchtmeetnet.nl/',
      }),
      dict({
        'logo': None,
        'name': 'Provincie Noord Brabant',
        'url': 'http://www.brabantluchtmeet.net/',
      }),
      dict({
        'logo': None,
        'name': 'World Air Quality Index Project',
        'url': 'https://waqi.info/',
      }),
    ]),
    'city': dict({
      'coordinates': Coordinates(latitude=51.6689355, longitude=4.541659),
      'external_url': 'https://aqicn.org/city/netherland/brabant/klundert',
      'location': None,
      'name': 'Klundert, Brabant',
    }),
    'dominant_pollutant': <Pollutant.PM10: 'pm10'>,
    'extended_air_quality': dict({
      'carbon_monoxide': None,
      'humidity': 71.6,
      'nitrogen_dioxide': 1.6,
      'ozone': None,
      'pm10': 4.0,
      'pm25': None,
      'pressure': 1018.0,
      'sulfur_dioxide': None,
      'temperature': 18.1,
    }),
    'measured_at': datetime.datetime(2023, 8, 7, 18, 0, tzinfo=datetime.timezone(datetime.timedelta(seconds=7200))),
    'station_id': 6337,
  })
# ---
# name: test_by_city[klundert]
  dict({
    'air_quality_index': 4,
    'attributions': list([
      dict({
        'logo': 'Netherland-RIVM.png',
        'name': 'RIVM - Rijksinstituut voor Volksgezondheid en Milieum, Landelijk Meetnet Luchtkwaliteit',
        'url': 'http://www.luchtmeetnet.nl/',
      }),
      dict({
        'logo': None,
        'name': 'Provincie Noord Brabant',
        'url': 'http://www.brabantluchtmeet.net/',
      }),
      dict({
        'logo': None,
        'name': 'World Air Quality Index Project',
        'url': 'https://waqi.info/',
      }),
    ]),
    'city': dict({
      'coordinates': Coordinates(latitude=51.6689355, longitude=4.541659),
      'external_url': 'https://aqicn.org/city/netherland/brabant/klundert',
      'location': None,
      'name': 'Klundert, Brabant',
    }),
    'dominant_pollutant': <Pollutant.PM10: 'pm10'>,
    'extended_air_quality': dict({
      'carbon_monoxide': None,
      'humidity': 71.6,
      'nitrogen_dioxide': 1.6,
      'ozone': None,
      'pm10': 4.0,
      'pm25': None,
      'pressure': 1018.0,
      'sulfur_dioxide': None,
      'temperature': 18.1,
    }),
    'measured_at': datetime.datetime(2023, 8, 7, 18, 0, tzinfo=datetime.timezone(datetime.timedelta(seconds=7200))),
    'station_id': 6337,
  })
# ---
# name: test_by_city[maarssen]
  dict({
    'air_quality_index': 29,
    'attributions': list([
      dict({
        'logo': 'Netherland-RIVM.png',
        'name': 'RIVM - Rijksinstituut voor Volksgezondheid en Milieum, Landelijk Meetnet Luchtkwaliteit',
        'url': 'http://www.luchtmeetnet.nl/',
      }),
      dict({
        'logo': None,
        'name': 'World Air Quality Index Project',
        'url': 'https://waqi.info/',
      }),
    ]),
    'city': dict({
      'coordinates': Coordinates(latitude=52.20153, longitude=4.987444),
      'external_url': 'https://aqicn.org/city/netherland/breukelen/snelweg',
      'location': None,
      'name': 'Snelweg, Breukelen',
    }),
    'dominant_pollutant': <Pollutant.OZONE: 'o3'>,
    'extended_air_quality': dict({
      'carbon_monoxide': None,
      'humidity': 77.5,
      'nitrogen_dioxide': 3.9,
      'ozone': 29.2,
      'pm10': 12.0,
      'pm25': 7.0,
      'pressure': 1008.4,
      'sulfur_dioxide': None,
      'temperature': 16.9,
    }),
    'measured_at': datetime.datetime(2023, 8, 7, 18, 0, tzinfo=datetime.timezone(datetime.timedelta(seconds=7200))),
    'station_id': 4586,
  })
# ---
# name: test_by_city[utrecht]
  dict({
    'air_quality_index': 29,
    'attributions': list([
      dict({
        'logo': 'Netherland-RIVM.png',
        'name': 'RIVM - Rijksinstituut voor Volksgezondheid en Milieum, Landelijk Meetnet Luchtkwaliteit',
        'url': 'http://www.luchtmeetnet.nl/',
      }),
      dict({
        'logo': None,
        'name': 'World Air Quality Index Project',
        'url': 'https://waqi.info/',
      }),
    ]),
    'city': dict({
      'coordinates': Coordinates(latitude=52.101308, longitude=5.128183),
      'external_url': 'https://aqicn.org/city/netherland/utrecht/griftpark',
      'location': None,
      'name': 'Griftpark, Utrecht',
    }),
    'dominant_pollutant': <Pollutant.OZONE: 'o3'>,
    'extended_air_quality': dict({
      'carbon_monoxide': None,
      'humidity': 73.2,
      'nitrogen_dioxide': 0.7,
      'ozone': 29.4,
      'pm10': 12.0,
      'pm25': 1.0,
      'pressure': 1008.8,
      'sulfur_dioxide': 0.1,
      'temperature': 17.5,
    }),
    'measured_at': datetime.datetime(2023, 8, 7, 17, 0, tzinfo=datetime.timezone(datetime.timedelta(seconds=7200))),
    'station_id': 6332,
  })
# ---
# name: test_get_by_coordinates
  dict({
    'air_quality_index': 29,
    'attributions': list([
      dict({
        'logo': 'Netherland-RIVM.png',
        'name': 'RIVM - Rijksinstituut voor Volksgezondheid en Milieum, Landelijk Meetnet Luchtkwaliteit',
        'url': 'http://www.luchtmeetnet.nl/',
      }),
      dict({
        'logo': None,
        'name': 'World Air Quality Index Project',
        'url': 'https://waqi.info/',
      }),
    ]),
    'city': dict({
      'coordinates': Coordinates(latitude=52.105031, longitude=5.124464),
      'external_url': 'https://aqicn.org/city/netherland/utrecht/de-jongweg',
      'location': None,
      'name': 'de Jongweg, Utrecht',
    }),
    'dominant_pollutant': <Pollutant.OZONE: 'o3'>,
    'extended_air_quality': dict({
      'carbon_monoxide': None,
      'humidity': 80.0,
      'nitrogen_dioxide': 2.3,
      'ozone': 29.4,
      'pm10': 12.0,
      'pm25': 17.0,
      'pressure': 1008.8,
      'sulfur_dioxide': None,
      'temperature': 16.0,
    }),
    'measured_at': datetime.datetime(2023, 8, 7, 17, 0, tzinfo=datetime.timezone(datetime.timedelta(seconds=7200))),
    'station_id': 4584,
  })
# ---
# name: test_get_by_ip
  dict({
    'air_quality_index': 26,
    'attributions': list([
      dict({
        'logo': 'Netherland-RIVM.png',
        'name': 'RIVM - Rijksinstituut voor Volksgezondheid en Milieum, Landelijk Meetnet Luchtkwaliteit',
        'url': 'http://www.luchtmeetnet.nl/',
      }),
      dict({
        'logo': None,
        'name': 'World Air Quality Index Project',
        'url': 'https://waqi.info/',
      }),
    ]),
    'city': dict({
      'coordinates': Coordinates(latitude=52.3702157, longitude=4.8951679),
      'external_url': 'https://aqicn.org/city/amsterdam',
      'location': None,
      'name': 'Amsterdam',
    }),
    'dominant_pollutant': <Pollutant.OZONE: 'o3'>,
    'extended_air_quality': dict({
      'carbon_monoxide': 2.3,
      'humidity': 69.5,
      'nitrogen_dioxide': 0.7,
      'ozone': 26.4,
      'pm10': 9.0,
      'pm25': 15.0,
      'pressure': 1016.4,
      'sulfur_dioxide': 0.2,
      'temperature': 17.5,
    }),
    'measured_at': datetime.datetime(2023, 8, 7, 17, 0, tzinfo=datetime.timezone(datetime.timedelta(seconds=7200))),
    'station_id': 5771,
  })
# ---
# name: test_get_by_name[failing_klundert]
  dict({
    'air_quality_index': None,
    'attributions': list([
      dict({
        'logo': 'Netherland-RIVM.png',
        'name': 'RIVM - Rijksinstituut voor Volksgezondheid en Milieum, Landelijk Meetnet Luchtkwaliteit',
        'url': 'http://www.luchtmeetnet.nl/',
      }),
      dict({
        'logo': None,
        'name': 'Provincie Noord Brabant',
        'url': 'http://www.brabantluchtmeet.net/',
      }),
      dict({
        'logo': None,
        'name': 'World Air Quality Index Project',
        'url': 'https://waqi.info/',
      }),
    ]),
    'city': dict({
      'coordinates': Coordinates(latitude=51.6689355, longitude=4.541659),
      'external_url': 'https://aqicn.org/city/netherland/brabant/klundert',
      'location': None,
      'name': 'Klundert, Brabant',
    }),
    'dominant_pollutant': <Pollutant.PM10: 'pm10'>,
    'extended_air_quality': dict({
      'carbon_monoxide': None,
      'humidity': 71.6,
      'nitrogen_dioxide': 2.0,
      'ozone': None,
      'pm10': 5.0,
      'pm25': None,
      'pressure': 1018.0,
      'sulfur_dioxide': None,
      'temperature': 18.1,
    }),
    'measured_at': datetime.datetime(2023, 8, 7, 19, 0, tzinfo=datetime.timezone(datetime.timedelta(seconds=7200))),
    'station_id': 6337,
  })
# ---
# name: test_get_by_name[klundert]
  dict({
    'air_quality_index': 5,
    'attributions': list([
      dict({
        'logo': 'Netherland-RIVM.png',
        'name': 'RIVM - Rijksinstituut voor Volksgezondheid en Milieum, Landelijk Meetnet Luchtkwaliteit',
        'url': 'http://www.luchtmeetnet.nl/',
      }),
      dict({
        'logo': None,
        'name': 'Provincie Noord Brabant',
        'url': 'http://www.brabantluchtmeet.net/',
      }),
      dict({
        'logo': None,
        'name': 'World Air Quality Index Project',
        'url': 'https://waqi.info/',
      }),
    ]),
    'city': dict({
      'coordinates': Coordinates(latitude=51.6689355, longitude=4.541659),
      'external_url': 'https://aqicn.org/city/netherland/brabant/klundert',
      'location': None,
      'name': 'Klundert, Brabant',
    }),
    'dominant_pollutant': <Pollutant.PM10: 'pm10'>,
    'extended_air_quality': dict({
      'carbon_monoxide': None,
      'humidity': 71.6,
      'nitrogen_dioxide': 2.0,
      'ozone': None,
      'pm10': 5.0,
      'pm25': None,
      'pressure': 1018.0,
      'sulfur_dioxide': None,
      'temperature': 18.1,
    }),
    'measured_at': datetime.datetime(2023, 8, 7, 19, 0, tzinfo=datetime.timezone(datetime.timedelta(seconds=7200))),
    'station_id': 6337,
  })
# ---
# name: test_get_by_station_number[372382]
  dict({
    'air_quality_index': 155,
    'attributions': list([
      dict({
        'logo': None,
        'name': 'Purple Air',
        'url': 'https://www.purpleair.com/',
      }),
      dict({
        'logo': None,
        'name': 'World Air Quality Index Project',
        'url': 'https://waqi.info/',
      }),
    ]),
    'city': dict({
      'coordinates': Coordinates(latitude=47.616673, longitude=-122.33384),
      'external_url': 'https://aqicn.org/station/@372382',
      'location': 'Terry Avenue, Belltown, Seattle, King County, Washington, 98101, USA',
      'name': 'Building Cure',
    }),
    'dominant_pollutant': <Pollutant.PM25: 'pm25'>,
    'extended_air_quality': dict({
      'carbon_monoxide': None,
      'humidity': 38.0,
      'nitrogen_dioxide': None,
      'ozone': None,
      'pm10': 59.0,
      'pm25': 155.0,
      'pressure': None,
      'sulfur_dioxide': None,
      'temperature': 25.5,
    }),
    'measured_at': datetime.datetime(2023, 8, 20, 19, 20, 22, tzinfo=datetime.timezone.utc),
    'station_id': -372382,
  })
# ---
# name: test_get_by_station_number[6337]
  dict({
    'air_quality_index': 5,
    'attributions': list([
      dict({
        'logo': 'Netherland-RIVM.png',
        'name': 'RIVM - Rijksinstituut voor Volksgezondheid en Milieum, Landelijk Meetnet Luchtkwaliteit',
        'url': 'http://www.luchtmeetnet.nl/',
      }),
      dict({
        'logo': None,
        'name': 'Provincie Noord Brabant',
        'url': 'http://www.brabantluchtmeet.net/',
      }),
      dict({
        'logo': None,
        'name': 'World Air Quality Index Project',
        'url': 'https://waqi.info/',
      }),
    ]),
    'city': dict({
      'coordinates': Coordinates(latitude=51.6689355, longitude=4.541659),
      'external_url': 'https://aqicn.org/city/netherland/brabant/klundert',
      'location': None,
      'name': 'Klundert, Brabant',
    }),
    'dominant_pollutant': <Pollutant.PM10: 'pm10'>,
    'extended_air_quality': dict({
      'carbon_monoxide': None,
      'humidity': 71.6,
      'nitrogen_dioxide': 2.0,
      'ozone': None,
      'pm10': 5.0,
      'pm25': None,
      'pressure': 1018.0,
      'sulfur_dioxide': None,
      'temperature': 18.1,
    }),
    'measured_at': datetime.datetime(2023, 8, 7, 19, 0, tzinfo=datetime.timezone(datetime.timedelta(seconds=7200))),
    'station_id': 6337,
  })
# ---
# name: test_search[failing_klundert]
  list([
    WAQISearchResult(air_quality_index=None, station_id=6337, station=Station(external_url='netherland/brabant/klundert', name='Klundert, Brabant', coordinates=Coordinates(latitude=51.6689355, longitude=4.541659))),
  ])
# ---
# name: test_search[klundert]
  list([
    WAQISearchResult(air_quality_index=4, station_id=6337, station=Station(external_url='netherland/brabant/klundert', name='Klundert, Brabant', coordinates=Coordinates(latitude=51.6689355, longitude=4.541659))),
  ])
# ---
# name: test_search[unknown]
  list([
  ])
# ---
